flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -197,19 +191,37 @@ public void onStart() throws Exception {
 			onFatalError(exception);
 			throw exception;
 		}
+
+		startRecoveredJobs();
```

Comment:
`onAddedJobGraph` submits the job graph from DispatcherProcess. Could we follow that pattern in DispatcherProcess start Dispatcher and follow by submit recovered job graphs? It would exclude "recovered" concern from `Dispatcher` and keep the recovery things out side of it.

> BTW `onAddedJobGraph` as well as `JobGraphStoreListener` is another ZK specific add-on high-availability concept.

Reply:
Well I notice that `submitJob` actually have a persisting step which isn't required for a recovered job. But if so, why we do this `onAddedJobGraph`...As is the code it works due to we support update job graph in job graph store.

Reply:
I would like to not change this behaviour. In the long run I would rather try to get rid of the `JobGraphListener` and the `onAddedJobGraph` and `onRemovedJobGraph` calls. Ideally, the leader recovers what it can recover at the beginning and then no more `JobGraphs` from the `JobGraphStore` will be accepted.

Yes, `submitJob` works for recovered jobs because the current `JobGraphWriter` implementations would simply overwrite the old value. Instead of introducing a new RPC method for the submission of recovered jobs, I would like to change this only if needed in the future. Moreover, I hope that we can get rid of the `JobGraphListener` interface even faster. Then this would not be needed.

Reply:
I agree. I am rearranging the design document attached in FLINK-10333 for emphasizing our proposal leader store based high-availability service, which includes the removal of `JobGraphListener` since it is originally introduced for patching coordinating concurrent modification.

I am going to start a discuss thread recently to see if our community is glad to accept it. Does it sound good to you? If so, what is a good time to start this thread?(I can finish the preparation in this week) I'm afraid that the community is blocked by many ongoing threads but IMO stability(high-availability) is critical component we'd better take more priority with.